File: app/src/main/java/ladsoft/com/popularmoviesapp/model/Movie.java
package ladsoft.com.popularmoviesapp.model;


import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.util.List;

import ladsoft.com.popularmoviesapp.BuildConfig;

public class Movie implements Parcelable {

    private static final String baseImageUrl = BuildConfig.BASE_IMAGE_URL;

    public Movie() {}

    public Movie(String posterPath, boolean adult, String overview, String releaseDate, List<Long> genreIds, long id, String originalTitle, String originalLanguage, String title, String backdropPath, double popularity, long voteCount, boolean video, int voteAverage, boolean favorite) {
        this.posterPath = posterPath;
        this.adult = adult;
        this.overview = overview;
        this.releaseDate = releaseDate;
        this.genreIds = genreIds;
        this.id = id;
        this.originalTitle = originalTitle;
        this.originalLanguage = originalLanguage;
        this.title = title;
        this.backdropPath = backdropPath;
        this.popularity = popularity;
        this.voteCount = voteCount;
        this.video = video;
        this.voteAverage = voteAverage;
        this.favorite = favorite;
    }

    @Expose @SerializedName("poster_path") private String posterPath;
    @Expose @SerializedName("is_adult") private boolean adult;
    @Expose @SerializedName("overview") private String overview;
    @Expose @SerializedName("release_date") private String releaseDate;
    @Expose @SerializedName("genre_ids") private List<Long> genreIds;
    @Expose @SerializedName("id") private long id;
    @Expose @SerializedName("original_title") private String originalTitle;
    @Expose @SerializedName("original_language") private String originalLanguage;
    @Expose @SerializedName("title") private String title;
    @Expose @SerializedName("backdrop_path") private String backdropPath;
    @Expose @SerializedName("popularity") private double popularity;
    @Expose @SerializedName("vote_count") private long voteCount;
    @Expose @SerializedName("video") private boolean video;
    @Expose @SerializedName("vote_average") private double voteAverage;
    private boolean favorite;

    protected Movie(Parcel in) {
        posterPath = in.readString();
        adult = in.readByte() != 0;
        overview = in.readString();
        releaseDate = in.readString();
        id = in.readLong();
        originalTitle = in.readString();
        originalLanguage = in.readString();
        title = in.readString();
        backdropPath = in.readString();
        popularity = in.readDouble();
        voteCount = in.readLong();
        video = in.readByte() != 0;
        voteAverage = in.readDouble();
        favorite = in.readByte() != 0;
//        in.readList(genreIds,  List.class.getClassLoader());
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(posterPath);
        dest.writeByte((byte) (adult ? 1 : 0));
        dest.writeString(overview);
        dest.writeString(releaseDate);
        dest.writeLong(id);
        dest.writeString(originalTitle);
        dest.writeString(originalLanguage);
        dest.writeString(title);
        dest.writeString(backdropPath);
        dest.writeDouble(popularity);
        dest.writeLong(voteCount);
        dest.writeByte((byte) (video ? 1 : 0));
        dest.writeDouble(voteAverage);
        dest.writeByte((byte) (favorite ? 1 : 0));
//        dest.writeList(genreIds);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<Movie> CREATOR = new Creator<Movie>() {
        @Override
        public Movie createFromParcel(Parcel in) {
            return new Movie(in);
        }

        @Override
        public Movie[] newArray(int size) {
            return new Movie[size];
        }
    };

    public String getPosterPath() {
        return baseImageUrl + posterPath;
    }

    public boolean isAdult() {
        return adult;
    }

    public String getOverview() {
        return overview;
    }

    public String getReleaseDate() {
        return releaseDate;
    }

    public List<Long> getGenreIds() { return genreIds; }

    public long getId() {
        return id;
    }

    public String getOriginalTitle() {
        return originalTitle;
    }

    public String getBackdropPath() {
        return backdropPath;
    }

    public double getPopularity() {
        return popularity;
    }

    public long getVoteCount() {
        return voteCount;
    }

    public boolean isVideo() {
        return video;
    }

    public double getVoteAverage() {
        return voteAverage;
    }

    public String getOriginalLanguage() {
        return originalLanguage;
    }

    public void setOriginalLanguage(String originalLanguage) {
        this.originalLanguage = originalLanguage;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public boolean isFavorite() {
        return favorite;
    }

    public void setFavorite(boolean favorite) {
        this.favorite = favorite;
    }
}

File: app/src/main/java/ladsoft/com/popularmoviesapp/model/Movie.kt
package ladsoft.com.popularmoviesapp.model


import android.database.Cursor
import android.os.Parcel
import android.os.Parcelable

import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName

import java.util.ArrayList

import ladsoft.com.popularmoviesapp.BuildConfig
import ladsoft.com.popularmoviesapp.data.MovieContract

data class Movie(@Expose @SerializedName("poster_path") var rawPosterPath: String = "",
    @Expose @SerializedName("is_adult") var isAdult: Boolean = false,
    @Expose @SerializedName("overview") var overview: String = "",
    @Expose @SerializedName("release_date") var releaseDate: String = "",
    @Expose @SerializedName("genre_ids") var genreIds: List<Long> = emptyList(),
    @Expose @SerializedName("id") var id: Long = 0.toLong(),
    @Expose @SerializedName("original_title") var originalTitle: String = "",
    @Expose @SerializedName("original_language") var originalLanguage: String = "",
    @Expose @SerializedName("title") var title: String? = null,
    @Expose @SerializedName("backdrop_path") var backdropPath: String = "",
    @Expose @SerializedName("popularity") var popularity: Double = 0.toDouble(),
    @Expose @SerializedName("vote_count") var voteCount: Long = 0.toLong(),
    @Expose @SerializedName("video") var isVideo: Boolean = false,
    @Expose @SerializedName("vote_average") var voteAverage: Double = 0.toDouble(),
    var isFavorite: Boolean = false) : Parcelable {

    constructor(cursor: Cursor) : this() {
        this.id = cursor.getLong(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_ID))
        this.rawPosterPath = cursor.getString(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_POSTER_PATH))
        this.isAdult = cursor.getInt(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_IS_ADULT)) == 1
        this.overview = cursor.getString(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_OVERVIEW))
        this.releaseDate = cursor.getString(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_RELEASE_DATE))
        this.genreIds = ArrayList<Long>()
        this.originalTitle = cursor.getString(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_ORIGINAL_TITLE))
        this.originalLanguage = cursor.getString(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_ORIGINAL_LANGUAGE))
        this.title = cursor.getString(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_TITLE))
        this.backdropPath = cursor.getString(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_BACKDROP_PATH))
        this.popularity = cursor.getDouble(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_POPULARITY))
        this.voteCount = cursor.getLong(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_VOTE_COUNT))
        this.isVideo = cursor.getInt(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_VIDEO)) == 1
        this.voteAverage = cursor.getDouble(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_VOTE_AVERAGE))
        this.isFavorite = cursor.getInt(cursor.getColumnIndex(MovieContract.MovieEntry.COLUMN_IS_FAVORITE)) == 1
    }

    protected constructor(`in`: Parcel) : this() {
        rawPosterPath = `in`.readString()
        isAdult = `in`.readByte().toInt() != 0
        overview = `in`.readString()
        releaseDate = `in`.readString()
        id = `in`.readLong()
        originalTitle = `in`.readString()
        originalLanguage = `in`.readString()
        title = `in`.readString()
        backdropPath = `in`.readString()
        popularity = `in`.readDouble()
        voteCount = `in`.readLong()
        isVideo = `in`.readByte().toInt() != 0
        voteAverage = `in`.readDouble()
        isFavorite = `in`.readByte().toInt() != 0
        //        in.readList(genreIds,  List.class.getClassLoader());
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(rawPosterPath)
        dest.writeByte((if (isAdult) 1 else 0).toByte())
        dest.writeString(overview)
        dest.writeString(releaseDate)
        dest.writeLong(id)
        dest.writeString(originalTitle)
        dest.writeString(originalLanguage)
        dest.writeString(title)
        dest.writeString(backdropPath)
        dest.writeDouble(popularity)
        dest.writeLong(voteCount)
        dest.writeByte((if (isVideo) 1 else 0).toByte())
        dest.writeDouble(voteAverage)
        dest.writeByte((if (isFavorite) 1 else 0).toByte())
        //        dest.writeList(genreIds);
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object {

        private val baseImageUrl = BuildConfig.BASE_IMAGE_URL

        @JvmField
        val CREATOR: Parcelable.Creator<Movie> = object : Parcelable.Creator<Movie> {
            override fun createFromParcel(`in`: Parcel): Movie {
                return Movie(`in`)
            }

            override fun newArray(size: Int): Array<Movie?> {
                return arrayOfNulls(size)
            }
        }
    }

    val posterPath: String
        get() = baseImageUrl + rawPosterPath!!
}
